Tools/AP_Periph/serial_tunnel.cpp

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+    if (monitor.uart == nullptr) {
+        return;
+    }
+    monitor.node_id = transfer->source_node_id;
+    monitor.protocol = pkt.protocol.protocol;
+    if (pkt.baudrate != monitor.baudrate) {
+        if (monitor.locked && pkt.baudrate != 0) {
+            monitor.uart->begin_locked(pkt.baudrate, 0, 0, TUNNEL_LOCK_KEY);
+            debug("begin_locked %u", unsigned(pkt.baudrate));
+        }
+        monitor.baudrate = pkt.baudrate;
+    }
+    monitor.uart->set_monitor_read_buffer(monitor.buffer);
+    monitor.last_request_ms = AP_HAL::millis();
+
+    // write to device
+    if (pkt.buffer.len > 0) {
+        if (monitor.locked) {
+            debug("write_locked %u", unsigned(pkt.buffer.len));
+            monitor.uart->write_locked(pkt.buffer.data, pkt.buffer.len, TUNNEL_LOCK_KEY);
+        } else {
+            monitor.uart->write(pkt.buffer.data, pkt.buffer.len);
+        }
+    } else {
+        debug("locked keepalive");
+    }
+}
+
+/*
+  send tunnelled serial data
+ */
+void AP_Periph_FW::send_serial_monitor_data()
+{
+    if (monitor.uart == nullptr ||
+        monitor.node_id == 0 ||
+        monitor.buffer == nullptr) {
+        return;
+    }
+    const uint32_t last_req_ms = monitor.last_request_ms;
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - last_req_ms >= 3000) {
+        // stop sending and unlock, but don't release the buffer
+        if (monitor.locked) {
+            debug("unlock");
+            monitor.uart->lock_port(0, 0);
+        }
+        monitor.uart = nullptr;
+        return;
+    }
+    if (monitor.locked) {
+        /*
+          when the port is locked nobody is reading the uart so the
+          monitor doesn't fill. We read here to ensure it fills
+         */
+        uint8_t buf[120];
+        for (uint8_t i=0; i<8; i++) {
+            if (monitor.uart->read_locked(buf, sizeof(buf), TUNNEL_LOCK_KEY) <= 0) {
+                break;
+            }
+        }
+    }
+    uint8_t sends = 8;
+    while (monitor.buffer->available() > 0 && sends-- > 0) {
+        uint32_t n;
+        const uint8_t *buf = monitor.buffer->readptr(n);
+        if (n == 0) {
+            return;
+        }
+        // broadcast data as tunnel packets, can be used for uCenter debug and device fw update
+        uavcan_tunnel_Targetted pkt {};
+        n = MIN(n, sizeof(pkt.buffer.data));
+        pkt.target_node = monitor.node_id;
+        pkt.protocol.protocol = monitor.protocol;
+        pkt.buffer.len = n;
+        pkt.baudrate = monitor.baudrate;
+        memcpy(pkt.buffer.data, buf, n);
+
+        uint8_t buffer[UAVCAN_TUNNEL_TARGETTED_MAX_SIZE] {};
+        const uint16_t total_size = uavcan_tunnel_Targetted_encode(&pkt, buffer, !canfdout());
+
+        debug("read %u", unsigned(n));
+
+        if (!canard_broadcast(UAVCAN_TUNNEL_TARGETTED_SIGNATURE,
+                              UAVCAN_TUNNEL_TARGETTED_ID,
+                              CANARD_TRANSFER_PRIORITY_MEDIUM,
```

Comment:
Should this be low? as this is intended as inflight feature as well.

Reply:
for reliable tunnelling for things like fw update I think we need it to be higher than LOW